crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+            || entropy == NULL || elen != ML_KEM_RANDOM_BYTES \
+            || (pub == NULL && prv == NULL)) \
+            return 0; \
+        return encap_seed(bits)(ctext, shared_secret, \
+                                constish(bits, pub, prv), entropy, ctx)
+
+    switch (v->rank) {
+    case_encap_seed(512);
+    case_encap_seed(768);
+    case_encap_seed(1024);
+    }
+    return 0;
+# undef case_encap_seed
+}
+
+int ossl_ml_kem_vencap_rand(vinfo_t v,
+                            uint8_t *ctext,
+                            size_t clen,
+                            uint8_t *shared_secret,
+                            size_t slen,
+                            const void *pub,
+                            void *prv,
+                            mctx *ctx)
+{
+# define case_encap_rand(bits) \
+    case ML_KEM_##bits##_RANK: \
+        if (ctext == NULL || clen != v->ctext_bytes \
+            || shared_secret == NULL \
+            || slen != ML_KEM_SHARED_SECRET_BYTES \
+            || (pub == NULL && prv == NULL)) \
+            return 0; \
+        return encap_rand(bits)(ctext, shared_secret, \
+                                constish(bits, pub, prv), ctx)
+
+    switch (v->rank) {
+    case_encap_rand(512);
+    case_encap_rand(768);
+    case_encap_rand(1024);
+    }
+    return 0;
+# undef case_encap_rand
+}
+
+int ossl_ml_kem_vdecap(vinfo_t v,
+                       uint8_t *shared_secret,
+                       size_t slen,
+                       const uint8_t *ctext,
+                       size_t clen,
+                       const void *prv,
+                       mctx *ctx)
+{
+# define case_decap(bits) \
+    case ML_KEM_##bits##_RANK: \
+        if (shared_secret == NULL || slen != ML_KEM_SHARED_SECRET_BYTES \
+            || ctext == NULL || prv == NULL) \
+            return 0; \
+        return decap(bits)(shared_secret, ctext, clen, prvcast(bits, prv), ctx)
+
+    switch (v->rank) {
+    case_decap(512);
+    case_decap(768);
+    case_decap(1024);
+    }
+    return 0;
+# undef case_decap
+}
+
+int ossl_ml_kem_vcompare_pubkeys(vinfo_t v1,
+                                 const void *pub1,
+                                 const void *prv1,
+                                 vinfo_t v2,
+                                 const void *pub2,
+                                 const void *prv2)
+{
+# define case_compare(bits) \
+    case ML_KEM_##bits##_RANK: \
+        /* No match if either or both are not available */ \
+        if (pub1 == NULL && prv1 == NULL) \
+            return 0; \
+        if (pub2 == NULL && prv2 == NULL) \
+            return 0; \
+        return memcmp(constish(bits, pub1, prv1)->pkhash, \
+                      constish(bits, pub2, prv2)->pkhash, \
+                      ML_KEM_PKHASH_BYTES) == 0
+
+    /*
+     * Rank mismatch should not happen, distinct ML-KEM variants have separate
+     * dispatch methods.
+     */
+    if (v1->rank != v2->rank)
```

Comment:
v1 != v2 would also work..

Reply:
I am now comparing public key SHA3-256 hashes, which are the same size across all variants. So any rank mismatch will also be detected, barring catastrophic failure of collision resistance in SHA3-256.

The more interesting question here is whether two "bare" keys (with just the "vinfo" and hash handles, no private or public key material as-yet generated or imported) are to be considered "equal". The ML-KEM implementation returns **no**, which makes sense to me, like SQL NULL values, we simply don't know whether the keys are really intended to be the same, and "no" seems like the safer answer.

No tests seem to expect equality of "bare" keys...